lib/api/src/sys/externals/memory.rs

```diff
@@ -172,61 +131,13 @@ impl Memory {
         self.handle.get_mut(store.objects_mut()).grow(delta.into())
     }
 
-    /// Safely reads bytes from the memory at the given offset.
-    ///
-    /// The full buffer will be filled, otherwise a `MemoryAccessError` is returned
-    /// to indicate an out-of-bounds access.
-    ///
-    /// This method is guaranteed to be safe (from the host side) in the face of
-    /// concurrent writes.
-    pub fn read(
-        &self,
-        store: &impl AsStoreRef,
-        offset: u64,
-        buf: &mut [u8],
-    ) -> Result<(), MemoryAccessError> {
-        self.buffer(store).read(offset, buf)
-    }
-
-    /// Safely reads bytes from the memory at the given offset.
-    ///
-    /// This method is similar to `read` but allows reading into an
-    /// uninitialized buffer. An initialized view of the buffer is returned.
-    ///
-    /// The full buffer will be filled, otherwise a `MemoryAccessError` is returned
-    /// to indicate an out-of-bounds access.
-    ///
-    /// This method is guaranteed to be safe (from the host side) in the face of
-    /// concurrent writes.
-    pub fn read_uninit<'a>(
-        &self,
-        store: &impl AsStoreRef,
-        offset: u64,
-        buf: &'a mut [MaybeUninit<u8>],
-    ) -> Result<&'a mut [u8], MemoryAccessError> {
-        self.buffer(store).read_uninit(offset, buf)
-    }
-
-    /// Safely writes bytes to the memory at the given offset.
-    ///
-    /// If the write exceeds the bounds of the memory then a `MemoryAccessError` is
-    /// returned.
-    ///
-    /// This method is guaranteed to be safe (from the host side) in the face of
-    /// concurrent reads/writes.
-    pub fn write(
-        &self,
-        store: &impl AsStoreRef,
-        offset: u64,
-        data: &[u8],
-    ) -> Result<(), MemoryAccessError> {
-        self.buffer(store).write(offset, data)
-    }
-
-    pub(crate) fn buffer<'a>(&'a self, store: &'a impl AsStoreRef) -> MemoryBuffer<'a> {
-        let definition = self.handle.get(store.as_store_ref().objects()).vmmemory();
-        let def = unsafe { definition.as_ref() };
-        MemoryBuffer {
+    /// In order to access the memory you must first lock it
+    /// (this will prevent it from growing)
+    pub fn lock<'a>(&'a self, store: &'a impl AsStoreRef) -> MemoryGuard<'a> {
+        let view = self.handle.get(store.as_store_ref().objects()).vmmemory();
+        let def = unsafe { view.as_ref() };
+        MemoryGuard {
+            _view: view,
             base: def.base,
             len: def.current_length,
             marker: PhantomData,
```

Comment:
I't like to name this `buffer` and return a `ArrayBuffer` (which is similar as a memory view) in case memory is not shared, and return a `SharedArrayBuffer` in case it's shared

Reply:
Similarly to JS:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer

Reply:
Sure no worries - makes sense

Reply:
this is no longer needed as the changes were simplified